### DISCUSSION TITLE
JavaScript actions node12 end-of-life

### DISCUSSION THREAD

**Discussion by MikeMcC399** [actions/runner#2604] [General] [answered]
https://github.blog/changelog/2023-05-04-github-actions-all-actions-will-run-on-node16-instead-of-node12/ says:

> Node 12 has been out of support since April 2022, as a result we have started the deprecation process of Node 12 for GitHub Actions. We plan to migrate all actions to run on Node16 by Summer 2023.
Following on from our warning in workflows using Node 12, we will start enforcing the use of Node16 rather than Node12 on the 18th of May.

Does "start enforcing" mean that all JavaScript workflows using `node12` will stop working on May 18, 2023 or is this a staged rollout process over weeks or months?

**Comment by ChristopherHX** [comment]
I think they mean all actions using `node12` will be forcefully executed by the `node16` runtime.

(There is a feature toggle in the public code to apply the change remotely)

Basically all node12 actions, which are not working with the node16 runtime will break.

> https://github.com/actions/runner/blob/49b04976f43be7d4122ba313db2193f303d1e8b7/src/Runner.Worker/Handlers/HandlerFactory.cs#L60-L64

**Reply by MikeMcC399** [reply]
@ChristopherHX 
Thank you very much for explaining how this works! PR https://github.com/actions/runner/pull/1716 gives a fuller picture of the change and shows the environment variable `ACTIONS_ALLOW_USE_UNSECURE_NODE_VERSION`.

**Reply by MikeMcC399** [reply]
Has this been enabled yet? Is there an easy way to check if a JavaScript action workflow has been forcefully updated to `node16`? (`ForceGithubJavascriptActionsToNode16`)

**Reply by ChristopherHX** [reply]
No, not enabled. Javascript actions still run on node12.
You can find the feature flags in the worker diag log.
```
  "variables": {
    "system.isScheduled": {
      "value": "False"
    },
    "system.servertype": {
      "value": "Hosted"
    },
    "system.culture": {
      "value": ""
    },
    "system.pipelineStartTime": {
      "value": "2023-05-23 14:53:38+00:00"
    },
    "system.runner.lowdiskspacethreshold": {
      "value": "100"
    },
    "DistributedTask.NewActionMetadata": {
      "value": "true"
    },
    "DistributedTask.EnableCompositeActions": {
      "value": "true"
    },
    "DistributedTask.EnhancedAnnotations": {
      "value": "true"
    },
    "DistributedTask.UploadStepSummary": {
      "value": "true"
    },
    "DistributedTask.AddWarningToNode12Action": {
      "value": "true"
    },
    "DistributedTask.AllowRunnerContainerHooks": {
      "value": "true"
    },
    "DistributedTask.DeprecateStepOutputCommands": {
      "value": "true"
    },
    "system.teamProjectId": {
      "value": "00000000-0000-0000-0000-000000000000"
    },
    "build.buildId": {
      "value": "0"
    },
    "build.containerId": {
      "value": "0"
    },
    "system.workflowFilePath": {
      "value": ".github/workflows/dispatch.yml"
    },
    "system.phaseDisplayName": {
      "value": "testa"
    },
    "system.phaseId": {
      "value": "00000000-0000-0000-0000-000000000000"
    },
    "system.phaseName": {
      "value": "testa"
    },
    "system.phaseAttempt": {
      "value": "1"
    },
    "system.jobIdentifier": {
      "value": "testa.Test_value"
    },
    "system.jobAttempt": {
      "value": "1"
    },
    "System.JobPositionInPhase": {
      "value": "1"
    },
    "System.TotalJobsInPhase": {
      "value": "1"
    },
    "system.github.job": {
      "value": "testa"
    },
    "system.jobDisplayName": {
      "value": "testa (Test value)"
    },
    "system.jobId": {
      "value": "00000000-0000-0000-0000-000000000000"
    },
    "system.jobName": {
      "value": "Test_value"
    },
    "system.github.token.permissions": {
      "value": "{\"Actions\":\"write\",\"Checks\":\"write\",\"Contents\":\"write\",\"Deployments\":\"write\",\"Discussions\":\"write\",\"Issues\":\"write\",\"Metadata\":\"read\",\"Packages\":\"write\",\"Pages\":\"write\",\"PullRequests\":\"write\",\"RepositoryProjects\":\"write\",\"SecurityEvents\":\"write\",\"Statuses\":\"write\"}"
    },
    "system.github.token": {
      "value": "***",
      "isSecret": true
    },
    "github_token": {
      "value": "***",
      "isSecret": true
    },
    "system.orchestrationId": {
      "value": "00000000-0000-0000-0000-000000000000.testa.test_value"
    },
    "system.runnerGroupName": {
      "value": "GitHub Actions"
    }
  },
  ```
There is the documentation about how to access the diag logs: https://docs.github.com/en/actions/monitoring-and-troubleshooting-workflows/enabling-debug-logging#enabling-runner-diagnostic-logging
The diag log is also enabled if you press rerun and enable the debug checkbox.

Or
```
- uses: actions/github-script@v5 # This is the last version with node12 runtime, this creates a job warning.
  with:
    script: console.log(process.version)
```

**Reply by MikeMcC399** [reply]
@ChristopherHX 
Thank you so much again for your advice! 👍🏻

**Comment by MikeMcC399** [comment]
The text of https://github.blog/changelog/2023-05-04-github-actions-all-actions-will-run-on-node16-instead-of-node12/ has been changed without there being a new announcement. It now says:

> Following on from our [warning in workflows using Node 12](https://github.blog/changelog/2022-09-22-github-actions-all-actions-will-begin-running-on-node16-instead-of-node12/), we will start enforcing the use of Node16 rather than Node12 on the 14th of June.

So the date has been moved from May 18, 2023 to June 14, 2023.

**Comment by MikeMcC399** [comment, marked as answer]
Today's blog entry June 13, 2023 provides more detailed information, see https://github.blog/changelog/2023-06-13-github-actions-all-actions-will-run-on-node16-instead-of-node12-by-default/

> Node12 has been out of support since [April 2022](https://github.com/nodejs/Release/#end-of-life-releases). As a result we have started the deprecation process of Node12 for GitHub Actions. We plan to migrate all actions to run on Node16 by Summer 2023.
Following on from our [warning in workflows using Node12](https://github.blog/changelog/2022-09-22-github-actions-all-actions-will-begin-running-on-node16-instead-of-node12/) we will start enforcing the use of Node16 rather than Node12 on the 14th of June.
>
> To opt out of this and continue using Node12 while it is still available in the runner, you can choose to set ACTIONS_ALLOW_USE_UNSECURE_NODE_VERSION=true
as an 'env' in their workflow or as an environment variable on your runner machine. This will only work until we upgrade the runner removing Node12 later in the summer.

**Reply by ChristopherHX** [reply]
The node12 warning was implicitly disabled with their force node 16 toggle, which has been enabled for me recently (some days after the changelog, first noticed today).

=> You don't even notice if you have been using an actions with node12 runtime on node16.

**Reply by MikeMcC399** [reply]
@ChristopherHX 
The warning about `node12` has been re-enabled, for instance:

> The following actions uses node12 which is deprecated and will be forced to run on node16: actions/checkout@v2. For more info: https://github.blog/changelog/2023-06-13-github-actions-all-actions-will-run-on-node16-instead-of-node12-by-default/

**Comment by MikeMcC399** [comment]
Comment also added to https://github.com/orgs/community/discussions/58226#discussioncomment-6230426

**Comment by ChristopherHX** [comment]
Removal of node12 planned (as of now 14th of August 2023) https://github.blog/changelog/2023-07-17-github-actions-removal-of-node12-from-the-actions-runner/.

This seem to indicate that node20 will come at a similar date or later, the azure pipelines team (https://github.com/microsoft/azure-pipelines-agent/pull/4346) just added node20 without removing old node versions.

**Reply by MikeMcC399** [reply]
I have asked again in https://github.com/orgs/community/discussions/58226#discussioncomment-6465852

**Comment by benwells** [comment]
Thank you all for compiling all of this information and for the engagement, sorry for the slow responses you all have been getting!  I know the node 12 removal changelog has been referenced here already. We do have work in flight in the coming weeks to address this node situation, and I put up a discussion post today that we plan to keep updated as we work.  Wanted to reference it here as well.  https://github.com/actions/runner/discussions/2704
